Introduce nkv package and nkvdriver

uncleDecart opened the pull request:
# Description

This PR completes the migration of EVE’s PubSub system to the Notify Key Value (nkv) architecture by integrating the nkv package, introducing the nkvdriver, and updating dependent services to comply with the new model.

It represents a major milestone in the PubSub restructuring effort — transitioning from the legacy socket-based peer-to-peer model to a centralized, protocol-driven key-value system.

# Summary of Changes

## Introduced the nkv Package

Added Notify Key Value (nkv) to the EVE ecosystem as the foundation for a protocol-based event and state synchronization system.
This forms the groundwork for consistent, cross-language messaging and decoupled service communication.

## Migrated PubSub to nkvdriver

Replaced the legacy socketdriver with the new nkvdriver, built on a centralized server model.

Key benefits:
- Eliminates redundant state sharing and improves topic consistency.
- Enables integration with non-Go services through a unified protocol.
- Centralizes observability with synchronized timestamps and structured logging.
- Reduces overall system complexity and maintenance overhead.
- “The king is dead, long live the king.”

## Updated Go Code to Modern Standards

Replaced all interface{} types with the modern any keyword for consistency with Go best practices.

## Standardized Topic Naming

Renamed DefaultVaultName from "Application Data Store" → "Application-Data-Store".
This resolves conflicts with nkv’s space-delimited topic format.
No migration logic needed since this publication is transient.

## Introduced Dedicated msrv PubSub Driver

The msrv service (running as a goroutine within zedagent) now uses its own driver.
nkvdriver enforces a single handler per topic per client, improving clarity and reliability.
This separation ensures msrv remains compatible with nkv without violating driver constraints.

## Refactored volumemgr for Safer Batch Processing

Replaced the handleRestart batching logic with generics.DeferredMapProcessor.

Reasons for change:

- The old socket-based approach relied on restart-signal semantics and guaranteed ordering — both incompatible with nkv.
- DeferredMapProcessor batches updates predictably and thread-safely.

Result: safer concurrency, cleaner logic, and alignment with nkv’s topic-based model.

# Motivation

The move to nkv addresses several key architectural limitations in the legacy PubSub system, as outlined in the ["Event-Driven Architecture and EVE"](https://docs.google.com/document/d/1PsKfe-4ngHKIGzQHIDo6zRXDwawFrwzZezQZtNAVm1M/edit?tab=t.0#heading=h.nyzxj89v3uw4)

# Size consideration

nkv-server for x64 is 1.4 Megabytes, nkv-client for debug container is 695 Kbytes. Splitting services into separate go processes would give us around 400 Megabytes overhead. With nkv, we can write mission critical services in separate processes using languages with better memory handling model (not garbage collector) such as Rust :crab:

**Important:** PubSub driver will be using [this](https://github.com/nkval/go-nkv) API, so it's good time to comment on the API and it's usage (see client app as an example)

For Rust library see [here](https://github.com/nkval/nkv/blob/main/src/lib.rs)

## Checklist

- [x] I've provided a proper description
- [ ] I've added the proper documentation (when applicable)
- [x] I've tested my PR on amd64 device(s)
- [ ] I've tested my PR on arm64 device(s)
- [ ] I've written the test verification instructions
- [x] I've set the proper labels to this PR

uncleDecart committed:
Introduce nkv package

This is step one of restructuring PubSub. This commit introduces
Notify Key Value (nkv) to EVE ecosystem. Next step would be to
use nkv as PubSub backend by introducing an nkv PubSub driver,
third and last step would be to switch PubSub to using nkv driver.
That way current systems won't change, but there will be opportunity
to switch to native nkv with easier API

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart commented:
Note that we need to create img in dockerhub, therefore current build is failing

rene commented:
@uncleDecart , is there a timeline for the upcoming PRs? Otherwise I would prefer to just merge this PR along with the other dependencies in place.... 

FYI: @eriknordmark , @OhmSpectator

deitch reviewed (approved):
This looks good, I really am glad to see this come in finally.

I have a few questions:

1. are we ready to merge this in? If nothing is using it, do we want another service up and running? Or will that happen really quickly?
2. Mounting `/persist` as a whole. Is that necessary? Will the entire database be in one single subdirectory of `/persist`, and so we can just mount that in?
3. The other services will connect to this via local network socket, or via Unix-domain socket, as implied by `--addr /run/nkv.sock` ? If Unix-domain, why does it need `net: host`?

I appreciate the responses, but none of that should hold this up.

uncleDecart commented:
>@uncleDecart , is there a timeline for the upcoming PRs? Otherwise I would prefer to just merge this PR along with the other dependencies in place....

@rene, my only concern is that there will be too much to review, otherwise I'm good with rolling it out together

rucoder commented:
@uncleDecart yes, as of now we need net to build rust. We are discussing with Avi how to avoid it and yet avoid gigantic vendor folder

uncleDecart committed:
pubsub: introduce nkvdriver and migrate from socketdriver

This commit introduces nkvdriver as a replacement for the legacy
socketdriver.

Unlike the previous peer-to-peer (p2p) model, nkvdriver uses a centralized
server to manage all publications and subscriptions. This shift enables
several improvements:

- Reduces overhead by avoiding redundant copying and simplifies reaching
consensus on specific topics.
- Enables support for new languages in EVE via a unified protocol and single
point of integration—no need to rely on filesystem-based state sharing.
- Centralizes visibility: all state changes and their processing can now be
observed from one place, with synchronized timestamps and consistent logs.

This marks a major architectural step toward cleaner, more maintainable, and
more extensible pubsub communication.

The king is dead, long live the king.

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
pubsub: change interface{} to any

to comply with modern golang style

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
pubsub: rename DefaultVaultName to Application-Data-Store

socketdriver allows spaces in topics, since nkv client server
protocol uses spaces to separate base64-encoded topics, it restricts
user from using spaces. The only topic, which had spaces was

vaultmgr.VaultConfig which used DefaultVaultName as part of key
for this topic.

This patch renames DefaultVaultName from "Applicationn Data Store" to
"Application-Data-Store" so that nkv can be used.

Note that since this publication is not persistent, there's no need
to handle renaming in upgradeconverter

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
pubsub: remove redundant parenthesis around handlers

because they're not needed

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
pubsub: introduce separate driver for msrv to support NKV constraints

msrv is a unique service that runs as a goroutine within zedagent.
Previously, it shared the same pubsub client, which worked fine with the
unixsocket p2p driver, as that driver allowed multiple subscriptions to
the same topic within a single client.

However, the new nkv driver enforces a stricter model: only one handler
per topic per client. This restriction improves code clarity,
debuggability, and overall structure—but it requires changes to existing
patterns.

To comply with this, msrv now uses a dedicated pubsub driver, ensuring
compatibility with nkv while maintaining separation of concerns.

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
feat: add DeferredMapProcessor[C] for time-based batching of deferred operations

This introduces a generic DeferredMapProcessor[C any] utility that enables
deferring the processing of map entries for a specified duration. It batches
frequent updates or creations and triggers a callback once the delay expires.

The processor is parameterized over the value type (C), with keys fixed as strings,
and is useful for reducing redundant work when items are rapidly added or updated.

Key features:
- Generic support with DeferredMapProcessor[C]
- Internal timer to trigger processing after delay
- Thread-safe LockedMap to store pending items
- Callback mechanism via LockedMapFunc[string, C]

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
volumemgr: replace handleRestart with generics.DeferredMapProcessor

volumemgr processes zedagent.VolumeConfig updates in a delayed, batched
manner. Previously, it collected configs in a map[string]*VolumeConfig and
processed them when a special restart signal was received on the same
topic. This approach relied on pubsub restart semantics and assumed reliable
ordering, which no longer holds for nkvdriver.

There are several issues with this method:

- The restart signal is not intended for batching. In multi-service
environments, this signal will have a different meaning and cannot safely
be reused this way.

- The use of a shared map introduces potential race conditions, since
handleModify can access the map concurrently, risking panics.

- The previous implementation relied on socketdriver's ordering guarantees:
objects sent on the same socket are received in the same sequence. For
example, sending:

VolumeConfig.UUID1, VolumeConfig.UUID2, VolumeConfig.__restart__

would ensure the same order of delivery and processing. nkvdriver,
however, uses a separate topic for restart signals, which breaks this
ordering assumption.

To address these issues, we now use generics.DeferredMapProcessor, which
batches config creation safely and predictably. When the first config arrives,
it defers processing for a fixed duration (15 seconds), allowing additional
configs to accumulate before processing them in batch.

This approach improves safety, aligns with nkv's design, and removes
dependency on restart signal semantics.

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
upgradeconverter: add movePersistPubsubToNkv for transitioning persisted pubsub state

To support upgrading from socketdriver to nkvdriver, we need to migrate
persisted pubsub subscriptions to a new storage structure compatible with NKV.

Previously, persisted subscriptions were stored under /persist/status/, but
this directory is shared by various services, making it difficult to
distinguish pubsub state from unrelated files.

The nkv driver introduces an ingest/digest pattern: incoming files are
first written to ingest/ and then atomically moved to digest/ once fully
received. This pattern requires both directories to reside on the same
filesystem to guarantee atomic operations.

As a result, we introduce /persist/nkv-storage/ingest/ and
/persist/nkv-storage/digest/ directories for pubsub state, and implement
movePersistPubsubToNkv to handle the migration.

Instead of a simple cp -r /persist/status/*, which risks copying unrelated
files, we selectively move known agent-specific directories:

/persist/status/<agentName> -> /persist/nkv-storage/digest/<agentName>

This ensures a clean migration path while preserving data integrity and
compatibility with the new storage model.

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
debug: bump nkv version container in debug container

to be consistent in hashes

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
pillar: update vendor

running go mod vendor

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
vendor trick to be discarded

Normally, this would be done in a follow-up PR after the change is merged and
available in the latest commit on github.com/lf-edge/eve. However, to avoid
breaking dependent services in the meantime, a temporary replace directive
can be added

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart committed:
gha: print docker verison for go tests

temp commit, will be removed

Signed-off-by: Pavel Abramov <uncle.decart@gmail.com>

uncleDecart commented:
go tests are passing locally, figuring out what's wrong with GitHub runners (for some reason docker socket is missing)

uncleDecart commented:
> > go tests are passing locally, figuring out what's wrong with GitHub runners (for some reason docker socket is missing)
> 
> The tests fail for me locally as well. Just start them with `make test`.

Fascinating, I ran those using go test

> I did not run docker in the tests, but containerd: https://github.com/lf-edge/eve/blob/master/pkg/pillar/agentlog/http-debug_test.go#L65

Any particular reason why you did it like that? I thought brining library to spin containers would be a way to go, maybe I missed something..

eriknordmark reviewed (commented):
There are scripts which read and write into the pubsub files.
One we talked about explicitly is in pkg/pillar/scripts, but a number of the eve-k shell scripts also read files. Two of them are
pkg/kube/pubsub.sh:ENCC_FILE_PATH="/persist/status/zedagent/EdgeNodeClusterConfig/global.json"
pkg/kube/cluster-init.sh:enc_status_file="/run/zedkube/EdgeNodeClusterStatus/global.json"

eriknordmark reviewed (commented):
I rebased this on master in my workspace and built it and put it on a device (with a ZFS /persist and with PCR enforcement) and I see the eveimage-update failing with
Watchdog report for IMGA EVE version 0.0.0-integrate-nkv-362c6a7f-dirty-2025-10-15.16.11-kvm-amd64 at 2025-10-15T19:30:46+00:00: 3 /run/zedbox.pid

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x365a525]

goroutine 4172 [running]:
github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.parseBaseOS(0xc0008c6008, 0xc0022b7b88?)
        /pillar/cmd/zedagent/parseconfig.go:326 +0x405
github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.parseConfig(0xc0008c6008, 0xc0022b7b88, 0x0)
        /pillar/cmd/zedagent/parseconfig.go:165 +0x7b4
github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.inhaleDeviceConfig(0xc0008c6008, 0xc0022b7b88, 0x0)
        /pillar/cmd/zedagent/handleconfig.go:1034 +0x2e5
github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.requestConfigByURL(0xc0008c6008, {0xc002134240, 0x5a}, 0x0, 0x0, 0x1, 0x0)
        /pillar/cmd/zedagent/handleconfig.go:810 +0x1267
github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.getLatestConfig(0xc0008c6008, 0x0, 0x1)
        /pillar/cmd/zedagent/handleconfig.go:868 +0x85
github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.configTimerTask(0xc0008c6008, 0xc001855810)
        /pillar/cmd/zedagent/handleconfig.go:392 +0x68
created by github.com/lf-edge/eve/pkg/pillar/cmd/zedagent.Run in goroutine 18
        /pillar/cmd/zedagent/zedagent.go:596 +0x1e7f

which is this line
	if status.ShortVersion != cfg.BaseOsVersion && cfg.Activate {

I don't know if you've changed something which can made status be nil from the previous line.

uncleDecart commented:
> > > I did not run docker in the tests, but containerd: https://github.com/lf-edge/eve/blob/master/pkg/pillar/agentlog/http-debug_test.go#L65
> > 
> > 
> > Any particular reason why you did it like that? I thought brining library to spin containers would be a way to go, maybe I missed something..
> 
> pillar kind of is a library to spin containers and I wanted to test it (and test that I can run a process in the debug container). Does this explain it?

Yes it does, thanks, will check if I can leverage that

europaul reviewed (commented):
just a couple of small things

rene reviewed (changes requested):
Blocking merge since somebody needs to takeover this......

rene commented:
Given the current state (so many conflicts), I'm closing this PR. Anyone taking this over, please, start from a fresh PR and link this one for visibility.
